[docs] Add a guide on how to generate a blurhash on a server (#20848)
expo/expo: docs/pages/versions/unversioned/sdk/image.mdx

```

<APISection packageName="expo-image" apiName="Image" />

## Generating a blurhash on a server

> Images can significantly improve the visual experience, but they can also slow down app/page loading times due to their large file sizes. To overcome this, we can create placeholder images using blurhash algorithm that can provide an immersive experience while deferring the loading of the actual picture until later.

Reviewer: ```suggestion
Images can significantly improve the visual experience, however, they can also slow down app/page loading times due to their large file sizes. To overcome this, you can create placeholder images using blurhash algorithm that provides an immersive experience while deferring the loading of the actual picture until later.
```

> This brief guide will demonstrate how to create a blurhash of an uploaded image on the backend using `JavaScript` and `Express.js`. However, the same techniques and principles apply to other languages and server technologies.

Reviewer: ```suggestion
This guide demonstrates how to create a blurhash of an uploaded image on the backend using JavaScript and Express.js. The same techniques and principles apply to other languages and server technologies.
```

> To simplify the process, we will install a few dependencies: [`multer`](https://github.com/expressjs/multer) for handling multipart requests, [`sharp`](https://github.com/lovell/sharp) for converting files to a data buffer, and the official [`blurhash JavaScript package`](https://github.com/woltapp/blurhash/tree/master/TypeScript).

Reviewer: ```suggestion
Start by installing a few dependencies: [`multer`](https://github.com/expressjs/multer) for handling multipart requests, [`sharp`](https://github.com/lovell/sharp) for converting files to a data buffer, and the official [`blurhash` JavaScript package](https://github.com/woltapp/blurhash/tree/master/TypeScript).
```

<Terminal cmd={['$ npm install multer sharp blurhash']} />

> Next, we need to import all needed functions from installed packages and initialized `multer`:

Reviewer: ```suggestion
Next, import all required functions from installed packages and initialize `multer`:
```

```js
// Multer is a middleware for handling `multipart/form-data`.
const multer = require('multer');
// Sharp allows you to recieve a data buffer from the uploaded image.
const sharp = require('sharp');
// Import the encode function from the blurhash package.
const { encode } = require('blurhash');

// Initilize `multer`.
const upload = multer();
```

> Assuming app is a variable that holds our express server, we can create an endpoint that accepts an image and returns a JSON response containing the generated blurhash.

Reviewer 2: ```suggestion
Assuming the `app` is a variable that holds our Express server, we can create an endpoint that accepts an image and returns a JSON response containing the generated blurhash.
```

Reviewer: Adding to the above suggestion, I'd avoid using second person voice (We, Our, etc.)

```js
app.post('/blurhash', upload.single('image'), async (req, res) => {
  const { file } = req;
  // If the file is not available we're returning with error.
  if (file === null) {
    res.status(400).json({ message: 'Image is missing' });
    return;
  }

>   // Users can specify number of components in each axies.

Reviewer 3: ```suggestion
  // Users can specify number of components in each axes.
```

"Axes" or "Axiis" is a plural - https://en.wiktionary.org/wiki/axis

>   const compoenentX = req.body.componentX ?? 4;
>   const compoenentY = req.body.componentY ?? 3;

Reviewer 2: ```suggestion
  const componentX = req.body.componentX ?? 4;
  const componentY = req.body.componentY ?? 3;
```

  // We're converting provided image to a byte buffer.

>   // Sharp currently supported multiple common formats like JPEG, PNG, WebP, GIF, and AVIF.

Reviewer 3: ```suggestion
  // Sharp currently supports multiple common formats like JPEG, PNG, WebP, GIF, and AVIF.
```

Wording nit.

>   // Sharp currencurrentlylty supported multiple common formats like JPEG, PNG, WebP, GIF, and AVIF.

Reviewer 2: > currencurrentlylty

What  ⁉️ 😂

  const { data, info } = await sharp(file.buffer).ensureAlpha().raw().toBuffer({
    resolveWithObject: true,
  });

>   const blurhash = encode(data, info.width, info.height, compoenentX, compoenentY);

Reviewer 2: ```suggestion
  const blurhash = encode(data, info.width, info.height, componentX, componentY);
```

  res.json({ blurhash });
});
```

> Additionally, the request can include two parameters: `componentX` and `componentY` which are passed through the algorithm. These values can be calculated or hard-coded on the server, or specified by the user. However, they must be within the range of 1 to 9 and have an aspect ratio similar to the uploaded image. A value of 9 will give the best results but may take longer to generate the hash.

Reviewer: ```suggestion
Additionally, the request can include two parameters: `componentX` and `componentY`, passed through the algorithm. These values can be calculated or hard-coded on the server or specified by the user. However, they must be within the range of 1 to 9 and have an aspect ratio similar to the uploaded image. A value of 9 will give the best results but may take longer to generate the hash.
```

> The process of generating a blurhash can be accomplished in various languages and server technologies, similar to the one using JavaScript. The key step is to locate an encoder for your chosen language, which can often be found in the [`woltapp/blurhash`](https://github.com/woltapp/blurhash#implementations) repository. Once you have the encoder, you will need to obtain a representation of the image. Some libraries use a default image class (for example, the Swift implementation uses `UIImage`), but in other cases, you will have to provide raw byte data. Therefore, be sure to check the encoder readme to confirm the expected data format.

Reviewer: ```suggestion
The process of generating a blurhash can be accomplished in various languages and server technologies, similar to the one using JavaScript. The key step is to locate an encoder for your chosen language, which can often be found in the [`woltapp/blurhash`](https://github.com/woltapp/blurhash#implementations) repository. Once you have the encoder, you will need to obtain a representation of the image. Some libraries use a default image class (for example, the Swift implementation uses `UIImage`). In other cases, you will have to provide raw byte data. Make sure to check the encoder's documentation to confirm the expected data format.
```

> > When working with raw byte data, please ensure the alpha layer is present (each pixel is represented by red, green, blue and alpha values). Failing to do so may lead to errors like "width and height must match the pixels array".

Reviewer: ```suggestion
> When working with raw byte data, ensure that the alpha layer is present (each pixel is represented by red, green, blue, and alpha values). Failing to do so will lead to errors such as "width and height must match the pixels array".
```

> > Note: when working with raw byte data, it is crucial to ensure that all pixels are represented by 4 numbers - the alpha layer is present. Failing to do so may lead to errors like "width and height must match the pixels array".

Reviewer: ```suggestion
> When working with raw byte data, it is crucial to ensure that all pixels are represented by 4 numbers - the alpha layer is present. Failing to do so may lead to errors such as "width and height must match the pixels array".
```

This phrase: "the alpha layer is present", doesn't make sense to me. Is the pixel composition represented by 4 numbers is called "the alpha layer"? If yes, we might want to rephrase the sentence and/or add a link, if possible, from the blurhash docs.
